### PR TITLE
Lind Strace

### PR DESCRIPTION
**Pull request by yizhuoliang**
## Description
Adding tracing feature at the interface between NaCl and the safeposix syscall implementations.
Thanks Rupesh and Yash for filling in the tracing functions @yashaswi2000 @rupeshkoushik07 
### Usage
- type `export TRACING_ENABLED=1` anywhere in the container shell
- make and install NaCl
- recompile the testing c file with the new NaCl
- tracing outputs goes to the strace_output.txt file at lind/lindenv/fs directory

### Type of change

- [ ] New feature (non-breaking change which adds functionality)

### Sample output with running socketselect.c
```
1 getuid() = -1
1 geteuid() = -1
1 getgid() = -1
1 getegid() = -1
1 open(./socketselect.nexe, 0, 0) = 3
1 read(3, 0x360ffffefa48, 832) = 832
1 lseek(3, 0x7f624a27ac58, 0) = 0
1 read(3, 0x360ffffef930, 36) = 36
1 fstat(3, 0x7f624a27ac00) = 0
1 getcwd(0x360f10040d38, 128) = 0
1 mmap(0x11000000, 196608, 0, 34, -1, 0x00000000) = 285212672
1 mmap(0x1000000, 65536, 5, 18, 3, 0x00010000) = -22
1 mmap((nil), 65536, 1, 2, 3, 0x00010000) = -16842752
1 munmap(0x360ffeff0000, 65536) = 0
1 mmap(0x11000000, 65536, 1, 2066, 3, 0x00020000) = 285212672
1 mmap(0x11010000, 65536, 3, 2066, 3, 0x00020000) = 285278208
1 mmap(0x11020000, 128, 3, 50, -1, 0x00000000) = 285343744
1 close(3) = 0
1 mutex_create() = 0
1 access(/etc/ld.so.preload, 4) = -2
1 open(/lib/glibc/libgcc_s.so.1, 0, 0) = 3
1 read(3, 0x360ffffef7d8, 832) = 832
1 lseek(3, 0x7f624a27ac58, 0) = 0
1 read(3, 0x360ffffef560, 36) = 36
1 fstat(3, 0x7f624a27ac00) = 0
1 mmap(0x11030000, 196608, 0, 34, -1, 0x00000000) = 285409280
1 mmap(0x1030000, 65536, 5, 18, 3, 0x00010000) = -22
1 mmap((nil), 65536, 1, 2, 3, 0x00010000) = -16842752
1 munmap(0x360ffeff0000, 65536) = 0
1 mmap(0x11030000, 65536, 1, 2066, 3, 0x00020000) = 285409280
1 mmap(0x11040000, 65536, 3, 2066, 3, 0x00020000) = 285474816
1 mmap(0x11050000, 212, 3, 50, -1, 0x00000000) = 285540352
1 close(3) = 0
1 open(/lib/glibc/libpthread.so.990e7c45, 0, 0) = 3
1 read(3, 0x360ffffef7b8, 832) = 832
1 lseek(3, 0x7f624a27ac58, 0) = 0
1 read(3, 0x360ffffef530, 36) = 36
1 fstat(3, 0x7f624a27ac00) = 0
1 mmap(0x11060000, 262144, 0, 34, -1, 0x00000000) = 285605888
1 mmap(0x1060000, 131072, 5, 18, 3, 0x00010000) = -22
1 mmap((nil), 131072, 1, 2, 3, 0x00010000) = -16908288
1 munmap(0x360ffefe0000, 131072) = 0
1 mmap(0x11060000, 65536, 1, 2066, 3, 0x00030000) = 285605888
1 mmap(0x11070000, 131072, 3, 2066, 3, 0x00030000) = 285671424
1 mmap(0x11090000, 8400, 3, 50, -1, 0x00000000) = 285802496
1 close(3) = 0
1 open(/lib/glibc/libc.so.990e7c45, 0, 0) = 3
1 read(3, 0x360ffffef798, 832) = 832
1 lseek(3, 0x7f624a27ac58, 0) = 0
1 read(3, 0x360ffffef520, 36) = 36
1 fstat(3, 0x7f624a27ac00) = 0
1 mmap(0x110a0000, 589824, 0, 34, -1, 0x00000000) = 285868032
1 mmap(0x10a0000, 1769472, 5, 18, 3, 0x00010000) = -22
1 mmap((nil), 1769472, 1, 2, 3, 0x00010000) = -18546688
1 munmap(0x360ffee50000, 1769472) = 0
1 mmap(0x110a0000, 393216, 1, 2066, 3, 0x001c0000) = 285868032
1 mmap(0x11100000, 131072, 3, 2066, 3, 0x00210000) = 286261248
1 mmap(0x11120000, 10128, 3, 50, -1, 0x00000000) = 286392320
1 close(3) = 0
1 mutex_lock(0) = 0
1 mutex_unlock(0) = 0
1 mmap((nil), 196608, 3, 131106, -1, 0x00000000) = -16973824
1 mmap((nil), 196608, 3, 131106, -1, 0x00000000) = -17170432
1 mutex_lock(0) = 0
1 cond_create() = 0
1 socket(2, 1, 0) = 3
1 setsockopt(3, 1, 2, 0x360ffefff310, 4) = 0
1 ioctl(3, 21537, 0x360ffefff310) = 0
1 socket(2, 1, 0) = 4
1 setsockopt(4, 1, 2, 0x360ffefcef1c, 4) = 0
1 mutex_lock(0) = 0
1 cond_create() = 1
1 bind(3, 0x360ffefff2b0, 16) = 0
1 listen(3, 1) = 0
1 mutex_lock(0) = 0
1 cond_signal(1) = 0
1 mutex_unlock(0) = 0
1 fstat(1, 0x7f624a269c00) = 0
1 mmap((nil), 8192, 3, 34, -1, 0x00000000) = -17235968
1 write(1, 0x360ffef90000, 23) = 23
1 cond_wait(1, 0) = 0
1 mutex_unlock(0) = 0
1 connect(4, 0x360ffefcf320, 16) = 0
1 send(4, 17, 0, 0x360f11000a70) = 17
1 select(4, 0x360ffefff1a0, 0x00000000, 0x00000000, 0x360ffefff2a0) = 1
1 write(1, 0x360ffef90000, 50) = 50
1 accept(3, 0x00000000, 0x00000000) = 5
1 write(1, 0x360ffef90000, 30) = 30
1 accept(3, 0x00000000, 0x00000000) = -11
1 write(1, 0x360ffef90000, 23) = 23
1 select(6, 0x360ffefff1a0, 0x00000000, 0x00000000, 0x360ffefff2a0) = 1
1 write(1, 0x360ffef90000, 27) = 27
1 recv(5, 80, 0, 0x360ffefff2c0) = 17
1 write(1, 0x360ffef90000, 20) = 20
1 send(5, 17, 0, 0x360ffefff2c0) = 17
1 mutex_lock(0) = 0
1 read(4, 0x360ffefcef20, 1024) = 17
1 cond_create() = 2
1 write(1, 0x360ffef90000, 18) = 18
1 close(4) = 0
1 mutex_lock(0) = 0
1 cond_signal(2) = 0
1 mutex_unlock(0) = 0
1 mutex_lock(0) = 0
1 cond_signal(0) = 0
1 mutex_unlock(0) = 0
1 cond_wait(0, 0) = 0
1 mutex_unlock(0) = 0
1 mutex_lock(0) = 0
1 cond_wait(2, 0) = 0
1 mutex_unlock(0) = 0
1 recv(5, 80, 0, 0x360ffefff2c0) = 0
1 close(5) = 0
1 write(1, 0x360ffef90000, 43) = 43
1 select(4, 0x360ffefff1a0, 0x00000000, 0x00000000, 0x360ffefff2a0) = 0
1 close(3) = 0
1 mutex_lock(0) = 0
1 cond_signal(0) = 0
1 mutex_unlock(0) = 0
1 cond_wait(0, 0) = 0
1 mutex_unlock(0) = 0
1 mutex_destroy(0) = 0
1 write(1, 0x360ffef90000, 36) = 36
```

### CODE REVIEW

**Comment by JustinCappos** [comment]
This is a great start!  We likely also want an option where we resolve arguments that are pointers to strings, ENUMs, etc. to more readable values.  However, that is a more complicated undertaking, so need not be in the first version of this.

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 1]
```diff
@@ -715,6 +748,9 @@ int32_t NaClSysRead(struct NaClAppThread  *natp,
 
   /* This cast is safe because we clamped count above.*/
   retval = (int32_t) read_result;
+  #ifdef TRACING
+  NaClStraceRead(nap->cage_id, d, (void *)sysaddr, count, retval);
+  #endif
```
Comment (rennergade) [inline src/trusted/service_runtime/nacl_syscall_common.c:762]: spacing

File: src/trusted/service_runtime/nacl_syscall_strace.c [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+}
+
+void NaClStraceRename(int cageid, const char *oldpath, const char *newpath, int32_t retval) {
+    fprintf(tracingOutputFile, "%d rename(oldpath: \"%s\", newpath: \"%s\") = %d\n", cageid, oldpath, newpath, retval);
+}
+void NaClStraceMmap(int cageid, void *start,size_t length,int prot,int flags,int d, uintptr_t offset, int32_t retval) {
+    fprintf(tracingOutputFile, "%d mmap(%p, %zu, %d, %d, %d, 0x%08"NACL_PRIxPTR") = %d\n", cageid, start, length, prot, flags, d, offset, retval);
+
+}
+void NaClStraceMunmap(int cageid, uintptr_t sysaddr, size_t length,int32_t retval) {
```
Comment (rennergade) [inline src/trusted/service_runtime/nacl_syscall_strace.c]: spacing

File: src/trusted/service_runtime/nacl_syscall_strace.c [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+   fprintf(tracingOutputFile, "%d munmap(0x%08"NACL_PRIxPTR", %zu) = %d\n", cageid, sysaddr, length, retval);
+
+}
```
Comment (rennergade) [inline src/trusted/service_runtime/nacl_syscall_strace.c:224]: spacing

**Review by rennergade — CHANGES_REQUESTED** [review]
I'll let others review before doing a more thorough review. My first comments are:

1. Remove the vscode file
2. Standardize spacing between the functions in the strace file and where the ifdefs are in syscall common. It'll make it easier to read. I'd rather err on the side of more newlines than making it compact.

File: src/trusted/service_runtime/build.scons [hunk 4]
```diff
@@ -21,6 +21,24 @@ if env.Bit('windows'):
...
+    env.Append(CCFLAGS=['-DTRACING=1'])
+    env.Append(CFLAGS=['-DTRACING=1'])
+    env.Append(CPPFLAGS=['-DTRACING=1'])
+    env.Append(CXXFLAGS=['-DTRACING=1'])
+
+#if os.environ.get('TRACING_TO_FILE') == '1':
+#    env.Append(CCFLAGS=['-DTRACING_TO_FILE=1'])
+#    env.Append(CFLAGS=['-DTRACING_TO_FILE=1'])
+#    env.Append(CPPFLAGS=['-DTRACING_TO_FILE=1'])
+#    env.Append(CXXFLAGS=['-DTRACING_TO_FILE=1'])
```
Comment (yzhang71) [inline src/trusted/service_runtime/build.scons]: Just a code style suggestion: if this part is commented out and will never be used, we can delete it.

File: src/trusted/service_runtime/nacl_syscall_strace.c [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+    FILE *newFile = fopen(path, "w");
+    if (newFile == NULL) {
+        perror("Error opening the tracing output file. Now output to stderr");
+        tracingOutputFile = stderr;
+    } else {
+        if (tracingOutputFile != stderr && tracingOutputFile != NULL) {
+            fclose(tracingOutputFile);
+        }
+        tracingOutputFile = newFile;
+        // setbuf(tracingOutputFile, NULL);
```
Comment (yzhang71) [inline src/trusted/service_runtime/nacl_syscall_strace.c]: Clean up

File: src/trusted/service_runtime/nacl_syscall_strace.c [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+}
+
+void NaClStraceSchedYield(int cageid, int ret) {
+    fprintf(tracingOutputFile, "%d schedyield() = %d\n", cageid, ret);
+}
+
+void NaClStraceExceptionHandler(int cageid, uint32_t             handler_addr,
+                                uint32_t             old_handler, int ret) {
+                                    fprintf(tracingOutputFile, "%d exceptionhandler(%u, %u) = %d\n", cageid, handler_addr, old_handler, ret);
+                                }
```
Comment (yzhang71) [inline src/trusted/service_runtime/nacl_syscall_strace.c]: The indentation needs to be fixed

File: src/trusted/service_runtime/nacl_syscall_strace.c [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+}
+
+void NaClStraceClockGetCommon(int cageid, int                   clk_id,
+                              uint32_t              ts_addr,
+                              uintptr_t            *time_func, 
+                              int ret) {
+                                fprintf(tracingOutputFile, "%d clockgetcommon(%d, %u, 0x%08"NACL_PRIxPTR") = %d\n",
+                                 cageid, clk_id, ts_addr, time_func, ret
+                                );
+                              }
```
Comment (yzhang71) [inline src/trusted/service_runtime/nacl_syscall_strace.c]: Indentation needs to be fixed

File: src/trusted/service_runtime/nacl_syscall_strace.c [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+    fprintf(tracingOutputFile, "%d mutex_lock(%d) = %d\n", cageid, mutex_handle, retval);
+
+}
+void NaClStraceMutexUnLock(int cageid, int32_t mutex_handle, int32_t retval) {
+    fprintf(tracingOutputFile, "%d mutex_unlock(%d) = %d\n", cageid, mutex_handle, retval);
+}
+void NaClStraceMutexTrylock(int cageid, int32_t mutex_handle, int32_t retval){
+    fprintf(tracingOutputFile, "%d mutex_trylock(%d) = %d\n", cageid, mutex_handle, retval);
+}
+void NaClStraceMutexDestroy(int cageid, int32_t mutex_handle,int32_t retval){
```
Comment (Yaxuan-w) [inline src/trusted/service_runtime/nacl_syscall_strace.c]: spacing before `int32_t`

File: src/trusted/service_runtime/nacl_syscall_strace.c (hunk 3 above)
Comment (yashaswi2000) [reply src/trusted/service_runtime/nacl_syscall_strace.c:224]: correcting it

File: src/trusted/service_runtime/nacl_syscall_strace.c (hunk 7 above)
Comment (yashaswi2000) [reply src/trusted/service_runtime/nacl_syscall_strace.c]: correcting

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 9]
```diff
@@ -3128,6 +3342,10 @@ int32_t NaClSysClockGetCommon(struct NaClAppThread  *natp,
   }
 
  done:
+  #ifdef TRACING
+  NaClStraceClockGetCommon(nap->cage_id, clk_id, ts_addr, (uintptr_t)time_func, retval);
```
Comment (yashaswi2000) [inline src/trusted/service_runtime/nacl_syscall_common.c:3435]: does this function need a trace call?

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 10]
```diff
@@ -3207,6 +3429,10 @@ int32_t NaClSysFork(struct NaClAppThread *natp) {
 
 fail:
 
+  #ifdef TRACING
+  NaClStraceFork(nap->cage_id, ret);
```
Comment (yashaswi2000) [inline src/trusted/service_runtime/nacl_syscall_common.c:3522]: this one too

File: src/trusted/service_runtime/nacl_syscall_common.c [hunk 11]
```diff
@@ -3628,6 +3863,10 @@ int32_t NaClSysWaitpid(struct NaClAppThread *natp,
   NaClLog(1, "[NaClSysWaitpid] options = %d \n", options);
   NaClLog(1, "[NaClSysWaitpid] ret = %d \n", ret);
 
+  #ifdef TRACING
+  NaClStraceWaitpid(nap->cage_id, pid, stat_loc, options, ret);
```
Comment (yashaswi2000) [inline src/trusted/service_runtime/nacl_syscall_common.c]: is this valid?

Comment (rennergade) [reply src/trusted/service_runtime/nacl_syscall_common.c]: stat loc is 32 bit, you want statlocptr i believe?

Comment (yizhuoliang) [reply src/trusted/service_runtime/nacl_syscall_common.c]: if you look at the syscall function, you see that `stat_loc` is a pointer passed into the syscall by the user, which is a user address, and there is a `sysaddr` variable declared to be the translated system address. I think using the sysaddr makes more sense here

Comment (yashaswi2000) [reply src/trusted/service_runtime/nacl_syscall_common.c]: there is no lind_* call happening inside it so i was confused

File: src/trusted/service_runtime/nacl_syscall_common.c (hunk 9 above)
Comment (yizhuoliang) [reply src/trusted/service_runtime/nacl_syscall_common.c:3435]: it seems like this is not an actual syscall, and this function is not even in the .h file, so it's a helper function, we don't want to trace this one

File: src/trusted/service_runtime/nacl_syscall_common.c (hunk 10 above)
Comment (yizhuoliang) [reply src/trusted/service_runtime/nacl_syscall_common.c:3522]: this is a valid (very important actually) syscall, we do want to trace this one

Comment (yashaswi2000) [reply src/trusted/service_runtime/nacl_syscall_common.c:3522]: making it

Comment (yashaswi2000) [reply src/trusted/service_runtime/nacl_syscall_common.c:3522]: this is being traced already, is the current structure correct?

**Review by rennergade — APPROVED** [review]
Looks great. Thanks guys!

**Pull request merged**